### PR TITLE
[SsrSite] feat: enable AWS_IAM auth on function URLs

### LINKED ISSUE #3090: [SsrSite] Lambda function URLs' auth is set to None
Related discussion on Discord: https://discord.com/channels/983865673656705025/1124011751973859438/1124011751973859438

# Issue

CloudFront can be bypassed by targeting the function URL directly.

This can be a problem if, for example:
- you want to enforce IP whitelisting via a CloudFront WAF,
- you want to protect (too) many functions from being maliciously provisionned (aka Denial of Wallet).

# Proposal

[] Enable the AWS_IAM authentication mechanism on function URLs
[] Configure a function@edge that signs CloudFront requests targeting a function URL

### PR DESCRIPTION
**Pull request by ealain**
Fixes #3090

### CODE REVIEW

**Comment by khuezy** [comment]
Thanks for putting this together! I skimmed over the PR and it looks good, but I'll do a thorough one later.

Question: How much over head would this introduce? Can this be an opt-in configuration? Or is the overhead insignificant and the AWS_IAM auth type standard/best practice that we should use it by default?

**Comment by ealain** [comment]
Hi! I think adding the opt-in configuration to this PR is a good idea.

In the future, when we have long-term real-world data, we can do a thorough benchmark on the performance and cost impact of the AWS_IAM auth type. Only then, we could create another PR to configure AWS_IAM by default.

**Comment by ealain** [comment]
Hi @khuezy!

@StanHannebelle implemented the opt-in configuration. By default, function URLs remain public.

I also fixed an issue where the signing function would not produce the correct signature on image optimization request because the `url` query param was already encoded by CloudFront.

Now the PR is ready for review.

File: packages/sst/src/constructs/SsrSite.ts [hunk 1]
```diff
@@ -983,10 +1009,12 @@ function handler(event) {
     const cfDistributionProps = cdk?.distribution || {};
 
     const fnUrl = this.serverLambdaForRegional!.addFunctionUrl({
-      authType: FunctionUrlAuthType.NONE,
+      authType: !this.props.enableIAMAuth
```
Comment (khuezy) [inline packages/sst/src/constructs/SsrSite.ts:1012]: Super nit: The other patterns above does a truth check, while this one uses `!` as the first ternary.

File: packages/sst/support/signing-function/helpers/queryParameterBagToQueryString.ts [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+ * This function does not encodeURIComponent the values,
+ * because the result is meant to be used in a JS object.
+ */
+export const queryParameterBagToQueryString = (
+  query: QueryParameterBag
+): string => {
+  const keys: Array<string> = [];
+  const serialized: Record<string, string> = {};
+
+  for (const key of Object.keys(query)) {
```
Comment (khuezy) [inline packages/sst/support/signing-function/helpers/queryParameterBagToQueryString.ts:13]: is `request.queryString` guaranteed to be an object, even if the url doesn't include it? Object.keys on null/undefined will throw an error.

**Review by khuezy — APPROVED** [review]
LGTM, some minor comments. @fwang PTAL

File: packages/sst/src/constructs/SsrSite.ts (hunk 1 above)
Comment (ealain) [reply packages/sst/src/constructs/SsrSite.ts]: Fixed

File: packages/sst/support/signing-function/helpers/queryParameterBagToQueryString.ts (hunk 2 above)
Comment (ealain) [reply packages/sst/support/signing-function/helpers/queryParameterBagToQueryString.ts:13]: Yes, this is guaranteed by the  `queryStringToQueryParameterBag` function where [`query` is initialized with `const query = {}`](https://github.com/ealain/sst/blob/67128b3c02c1d5e65a3ea167551330186be7016b/packages/sst/support/signing-function/helpers/queryStringToQueryParameterBag.ts#L6).

**Comment by luc-ferron** [comment]
I think a few unit tests to validate the feature is working, and that it doesn't get broken by others in the future, would be a good idea. Thanks for putting all this together, I'm going to remove my home made support of IAM in favor of this when it gets in.

**Comment by ealain** [comment]
@luc-ferron thank you for your support!

I agree that tests will definitely help the project in the future 🚀

However, I think adding tests in this PR is not appropriate because there is no testing framework already set up. This topic should be addressed by the core team.

**Comment by luc-ferron** [comment]
> @luc-ferron thank you for your support!
> 
> I agree that tests will definitely help the project in the future 🚀
> 
> However, I think adding tests in this PR is not appropriate because there is no testing framework already set up. This topic should be addressed by the core team.

I do see a lot of tests in this folder for each high level construct: 

https://github.com/serverless-stack/sst/tree/master/packages/sst/test/constructs

Adding some validation of the additional feature in NextJSSite that inherits SsrSite could be a quick option?

**Comment by ealain** [comment]
I just changed the dependency @aws-sdk/signature-v4 which was moved to @smithy/signature-v4 a few days ago 😄 

@jayair @fwang What are your thoughts on what should be tested here?

**Comment by jayair** [comment]
Yup let me check on it with Frank.

**Comment by fwang** [comment]
Made some cosmetic changes and added unit tests.

Thanks @ealain @StanHannebelle!

**Comment by khuezy** [comment]
🙏 How much latency does it actually add?

**Comment by fwang** [comment]
From my brief testing.. it seems the signing function takes ~200 to boot up (cold start) and <20ms to run.

**Comment by khuezy** [comment]
Not bad on the warm run, can we warm up the signing function?

**Comment by khuezy** [comment]
Actually, would it be possible to implement the signing function in rust or something to minimize the coldstart?

**Pull request merged**